Samples/BackgroundBlur/BackgroundBlur/TransformAsync.cpp

```diff
@@ -0,0 +1,1212 @@
+            ** we know m_spEventQueue can never be
+            ** NULL due to InitializeTransform()
+            ***************************************/
+            hr = m_spEventQueue->QueueEvent(pDrainCompleteEvent.get());
+            if (FAILED(hr))
+            {
+                break;
+            }
+        }
+        else
+        {
+            m_dwStatus |= (MYMFT_STATUS_DRAINING);
+
+        }
+        /*******************************
+        ** Note: This MFT only has one
+        ** input stream, so it does not
+        ** track the stream being drained.
+        ** If your MFT has more than one
+        ** input stream, you will
+        ** have to change this logic
+        *******************************/
+    } while (false);
+
+
+    return hr;
+}
+
+HRESULT TransformAsync::OnFlush(void)
+{
+    HRESULT hr = S_OK;
+    do
+    {
+        // TODO: Remove fence values so that framerate can restart
+        std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+        m_dwStatus &= (~MYMFT_STATUS_STREAM_STARTED);
+
+        hr = FlushSamples();
+        if (FAILED(hr))
+        {
+            break;
+        }
+
+    } while (false);
+
+    return hr;
+}
+
+HRESULT TransformAsync::OnMarker(
+    const ULONG_PTR pulID)
+{
+    HRESULT hr = S_OK;
+
+    do
+    {
+        // No need to lock, our sample queue is thread safe
+
+        /***************************************
+        ** Since this in an internal function
+        ** we know m_pInputSampleQueue can never be
+        ** NULL due to InitializeTransform()
+        ***************************************/
+
+        hr = m_pInputSampleQueue->MarkerNextSample(pulID);
+        if (FAILED(hr))
+        {
+            break;
+        }
+    } while (false);
+
+
+    return hr;
+}
+
+HRESULT TransformAsync::RequestSample(
+    const UINT32 un32StreamID)
+{
+    HRESULT         hr = S_OK;
+    com_ptr<IMFMediaEvent> pEvent;
+
+    do
+    {
+        {
+            std::lock_guard<std::recursive_mutex> lock(m_mutex);
+
+            if ((m_dwStatus & MYMFT_STATUS_STREAM_STARTED) == 0)
+            {
+                // Stream hasn't started
+                hr = MF_E_NOTACCEPTING;
+                break;
+            }
+        }
+
+        hr = MFCreateMediaEvent(METransformNeedInput, GUID_NULL, S_OK, NULL, pEvent.put());
+        if (FAILED(hr))
+        {
+            break;
+        }
+        if (pEvent == NULL)
+        {
+            TRACE((L"RequestSample: pEvent null not queueing anything!"));
+        }
+
+        hr = pEvent->SetUINT32(MF_EVENT_MFT_INPUT_STREAM_ID, un32StreamID);
+        if (FAILED(hr))
+        {
+            break;
+        }
+
+        /***************************************
+        ** Since this in an internal function
+        ** we know m_spEventQueue can never be
+        ** NULL due to InitializeTransform()
+        ***************************************/
+
+        {
+            std::lock_guard<std::recursive_mutex> lock(m_mutex);
```

Comment:
why are you reaquiring the lock create event and set property should be fast so no need to unlock